// @solana-program/memo/dist/src/index.js
'use strict';

var web3_js = require('@solana/web3.js');

// src/generated/instructions/addMemo.ts

// src/generated/programs/memo.ts
var MEMO_PROGRAM_ADDRESS = "MemoSq4gqABAXKb96qnH8TysNcWxMyWCqXgDLGmfcHr";
var MemoInstruction = /* @__PURE__ */ ((MemoInstruction2) => {
  MemoInstruction2[MemoInstruction2["AddMemo"] = 0] = "AddMemo";
  return MemoInstruction2;
})(MemoInstruction || {});

// src/generated/instructions/addMemo.ts
function getAddMemoInstructionDataEncoder() {
  return web3_js.getStructEncoder([["memo", web3_js.getUtf8Encoder()]]);
}
function getAddMemoInstructionDataDecoder() {
  return web3_js.getStructDecoder([["memo", web3_js.getUtf8Decoder()]]);
}
function getAddMemoInstructionDataCodec() {
  return web3_js.combineCodec(
    getAddMemoInstructionDataEncoder(),
    getAddMemoInstructionDataDecoder()
  );
}
function getAddMemoInstruction(input, config) {
  const programAddress = config?.programAddress ?? MEMO_PROGRAM_ADDRESS;
  const args = { ...input };
  const remainingAccounts = (args.signers ?? []).map(
    (signer) => ({
      address: signer.address,
      role: web3_js.AccountRole.READONLY_SIGNER,
      signer
    })
  );
  const instruction = {
    accounts: remainingAccounts,
    programAddress,
    data: getAddMemoInstructionDataEncoder().encode(
      args
    )
  };
  return instruction;
}
function parseAddMemoInstruction(instruction) {
  return {
    programAddress: instruction.programAddress,
    data: getAddMemoInstructionDataDecoder().decode(instruction.data)
  };
}

exports.MEMO_PROGRAM_ADDRESS = MEMO_PROGRAM_ADDRESS;
exports.MemoInstruction = MemoInstruction;
exports.getAddMemoInstruction = getAddMemoInstruction;
exports.getAddMemoInstructionDataCodec = getAddMemoInstructionDataCodec;
exports.getAddMemoInstructionDataDecoder = getAddMemoInstructionDataDecoder;
exports.getAddMemoInstructionDataEncoder = getAddMemoInstructionDataEncoder;
exports.parseAddMemoInstruction = parseAddMemoInstruction;
